src/pykeen/contrib/lightning.py

```diff
@@ -0,0 +1,130 @@
+
+
+class LitLCWAModule(pytorch_lightning.LightningModule):
+    """A PyTorch Lightning module for training a model with LCWA training loop."""
+
+    def __init__(
+        self,
+        # dataset
+        dataset: HintOrType[Dataset] = "nations",
+        dataset_kwargs: OptionalKwargs = None,
+        # model
+        model: HintOrType[Model] = "distmult",
+        model_kwargs: OptionalKwargs = None,
+        # stored outside of the training loop / optimizer to give access to auto-tuning from Lightning
+        batch_size: int = 32,
+        learning_rate: float = 1.0e-03,
+        # optimizer
+        optimizer: HintOrType[torch.optim.Optimizer] = None,
+        optimizer_kwargs: OptionalKwargs = None,
+    ):
+        """
+        Create the lightning module.
+
+        :param dataset:
+            the dataset, or a hint thereof
+        :param dataset_kwargs:
+            additional keyword-based parameters passed to the dataset
+
+        :param model:
+            the model, or a hint thereof
+        :param model_kwargs:
+            additional keyword-based parameters passed to the model
+
+        :param batch_size:
+            the training batch size
+        :param learning_rate:
+            the learning rate
+
+        :param optimizer:
+            the optimizer, or a hint thereof
+        :param optimizer_kwargs:
+            additional keyword-based parameters passed to the optimizer. should not contain `lr`, or `params`.
+        """
+        super().__init__()
+        self.dataset = get_dataset(dataset=dataset, dataset_kwargs=dataset_kwargs)
+        self.model = model_resolver.make(model, model_kwargs, triples_factory=self.dataset.training)
+        self.loss = self.model.loss
+        self.optimizer = optimizer
+        self.optimizer_kwargs = optimizer_kwargs
+        self.learning_rate = learning_rate
+        self.batch_size = batch_size
+
+    def forward(self, x):
+        """
+        Perform the prediction or inference step.
+
+        .. note::
+            in lightning, forward defines the prediction/inference actions
+        """
+        return self.model.predict_t(x)
+
+    def _step(self, batch, prefix: str):
+        """Refactored step."""
+        hr_batch, labels = batch
+        scores = self.model.score_t(hr_batch=hr_batch)
+        loss = self.loss.process_lcwa_scores(predictions=scores, labels=labels)
+        self.log(f"{prefix}_loss", loss)
+        return loss
+
+    def training_step(self, batch, batch_idx):
+        """Perform a training step."""
+        return self._step(batch, prefix="train")
+
+    def validation_step(self, batch, batch_idx, *args, **kwargs):
+        """Perform a validation step."""
+        return self._step(batch, prefix="val")
+
+    def _dataloader(self, triples_factory: CoreTriplesFactory, shuffle: bool = False) -> torch.utils.data.DataLoader:
+        """Create a data loader."""
+        return torch.utils.data.DataLoader(
```

Comment:
can we put functionality for creating a dataloader in the triples factory itself? Seems pretty common of an operation

Reply:
we currently have such functionality in the training loops, cf.
https://github.com/pykeen/pykeen/blob/f81b583a13761b83c1dfe537ee48d23368f75982/src/pykeen/training/training_loop.py#L776-L805

Reply:
so then this could be deduplicated there? Maybe better for a different PR later, so you can be as flexible as you want to get this one working